### PR TITLE
modify _expand_bounds() shape checks to work with >2-dim bounds.

### PR DESCRIPTION
**Pull request by saitcakmak**
## Motivation

`botorch.optim.utils._expand_bounds()` raises an error when used with `bounds` with more than 2 dimensions. This PR modifies the shape checks in `_expand_bounds()` to allow more general shapes, as long as `bounds` has the same t-batch shape as `X`.

This change extends the functionality of `gen_candidates_scipy` to allow optimization with a custom bound for each entry of `initial_conditions`.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/pytorch/botorch/blob/master/CONTRIBUTING.md#pull-requests)?

Yes.

## Test Plan

Added unit tests.

### CODE REVIEW

File: test/optim/test_utils.py [hunk 1]
```diff
@@ -153,7 +169,7 @@ def test_get_extra_mll_args(self):
 
 class TestExpandBounds(BotorchTestCase):
     def test_expand_bounds(self):
-        X = torch.zeros(2, 3)
+        X = torch.zeros(4, 2, 3)
```
Comment (Balandat) [inline test/optim/test_utils.py:172]: we should probably test with and without explicit t-batch dimension

File: botorch/optim/utils.py [hunk 2]
```diff
@@ -153,7 +153,11 @@ def _expand_bounds(
             ebounds = bounds.view(1, -1)
         else:
             ebounds = bounds
-        if ebounds.shape[1] != X.shape[-1]:
+        if (
+            ebounds.shape[-1] != X.shape[-1]
+            or ebounds.shape[-2] not in [1, X.shape[-2]]
+            or (ebounds.dim() > 2 and ebounds.shape[:-2] != X.shape[:-2])
```
Comment (Balandat) [inline botorch/optim/utils.py:159]: So this requires bounds to be defined with explicit t-batch dim. An alternative would be to actually fully support broadcasting - would that make sense her? I wrote some utilities for this a while ago that may be helpful here: https://github.com/cornellius-gp/gpytorch/blob/master/gpytorch/utils/broadcasting.py#L6

**Review by Balandat — COMMENTED** [review]
Cool, thanks for the PR. 

What's the specific motivation for optimizing different elements w.r.t. different bounds? Is this to deal e.g. with different fidelities in look-ahead approaches?

File: test/optim/test_utils.py (hunk 1 above)
Comment (saitcakmak) [reply test/optim/test_utils.py:172]: Just to clarify, do we want to test without explicit t-batch when `bounds` is <=2 dim, or even when `bounds` has a t-batch? If it's the second case, I am assuming this should raise an error if `bounds` has a t-batch and `X` doesn't.

File: botorch/optim/utils.py (hunk 2 above)
Comment (saitcakmak) [reply botorch/optim/utils.py:159]: I looked into the behavior of `Tensor.max()` (or `min()`), and it is safe as long as `bounds` is broadcastable to `X`. It matches the expected behavior as well, so that should work here. I will look into a simpler condition for this.

**Comment by saitcakmak** [comment]
Thanks for the quick review!

> What's the specific motivation for optimizing different elements w.r.t. different bounds?

I was implementing batch BnB, using `gen_candidates_scipy` to optimize relaxations. Naturally, each branch is optimized over a different solution space. This happened to be the only thing that raised an issue.

File: botorch/optim/utils.py (hunk 2 above)
Comment (Balandat) [reply botorch/optim/utils.py:159]: You may be able to just use the broadcasting utils directly? Or I guess that doesn't cover the case when `bounds` has explicit t-batch dims but `X` doesn't?

**Comment by Balandat** [comment]
> I was implementing batch BnB, using gen_candidates_scipy to optimize relaxations.

Cool, excited to see how this works.

**Comment by saitcakmak** [comment]
I modified it to use `bounds.expand_as(X)`. This quite general and has the correct behavior. According to PyTorch documentation, `expand` does not create new memory, so I don't see any downside to this.

This does not handle the case when `bounds` has explicit t-batch but `X` doesn't. I verified that `_expand_bounds` is only used within `columnwise_clamp`, which is only used in `gen_candidates_*`. I don't see any use case where `bounds` would have t-batch and `X` wouldn't, so I think it is safe to ignore it. This will raise an error in that case.

**Review by Balandat — APPROVED** [review]
great, this lgtm

**Comment by Balandat** [comment]
Thanks for contributing.
